[Merged by Bors] - feat(number_theory/lucas_primality): Add theorem for Lucas primality test

BoltonBailey opened the pull request:
This is a PR for adding the [Lucas primality test](https://en.wikipedia.org/wiki/Lucas_primality_test) to mathlib. This tells us that a number `p` is prime when an element `a : zmod p` has order `p-1` .

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #8833 
- [x] depends on: #8842 

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

BoltonBailey committed:
totient todos

BoltonBailey committed:
work on main theorem

Ruben-VandeVelde committed:
feat(data/nat/totient): add nat.totient_prime_iff

Ruben-VandeVelde commented on src/data/nat/totient.lean:
```diff
@@ -172,6 +206,20 @@ by rcases exists_eq_succ_of_ne_zero (pos_iff_ne_zero.1 hn) with ⟨m, rfl⟩;
+  sorry,
+end
+
+lemma totient_prime_iff {p : ℕ} : p.prime ↔ φ p = p - 1 :=
```
I PR'd this as #8833

BoltonBailey replied:
Excellent, thank you. I have merged your `totient_prime_iff` branch into this one.

ericrbg replied:
you should probably use the depend bot thing

BoltonBailey replied:
I don't know what you mean by "the depend bot thing", sorry.

ericrbg replied:
see what Yakov did in #8859, for example. you put the PRs you need in your description, and it won't let you merge until they are merged

BoltonBailey replied:
Oh, ha ha, I thought you were saying something about \bot the logical concept.

BoltonBailey committed:
Merge branch 'totient-prime-iff' of github.com:leanprover-community/mathlib into BoltonBailey/lucas-primality

pechersky commented on src/data/nat/totient.lean:
```diff
@@ -39,6 +39,40 @@ lemma totient_le (n : ℕ) : φ n ≤ n :=
            ... = n              : card_range _
 
+-- TODO move next to filter_subset_filter
+lemma filter_subset_filter_of_implies (s : finset ℕ) (p q : ℕ → Prop) [decidable_pred p] [decidable_pred q] (h : ∀ n , p n -> q n) : s.filter p ⊆ s.filter q :=
```
Do we have a lemma of the form `monotone s.filter` for any list s over any type?

BoltonBailey replied:
Good point that this is an instance of monotonicity. I was not able to find such a lemma just now.

pechersky replied:
#8842

BoltonBailey committed:
Add Eric Rodriguez's code finishing prime_iff_card_units

BoltonBailey committed:
add zmod_order_le_card_units

BoltonBailey committed:
progress on main theorem

BoltonBailey committed:
todo to remove

BoltonBailey committed:
small progess

BoltonBailey committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into BoltonBailey/lucas-primality

BoltonBailey committed:
remove monotonicity lemma

BoltonBailey committed:
refactor to use fact (0 < p)

BoltonBailey committed:
remove extraneous old code

BoltonBailey committed:
finished totient_lt

BoltonBailey committed:
finished proof

BoltonBailey committed:
fix line too long

Ruben-VandeVelde commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,215 @@
+      rw nat.totient_eq_iff_prime at h,
+      exact h,
+      exact hp.1, },
+end
```
This can be golfed to:
```suggestion
lemma prime_iff_card_units (p : ℕ) [fact (0 < p)] :
  p.prime ↔ fintype.card (units (zmod p)) = p - 1 :=
by rw [zmod.card_units_eq_totient, nat.totient_eq_iff_prime (fact.out (0 < p))]
```

and should probably move to `src/data/nat/totient.lean` (along with the previous lemma).

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

BoltonBailey committed:
delete scratch work

BoltonBailey committed:
refactor dvd_div_iff_mul_dvd

BoltonBailey committed:
split off lemma from theorem

BoltonBailey committed:
golf

BoltonBailey committed:
fix

BoltonBailey committed:
golfing

BoltonBailey committed:
formatting

BoltonBailey committed:
formatting

BoltonBailey committed:
golf

BoltonBailey committed:
Move lemma to divisibility file

BoltonBailey committed:
fix

BoltonBailey committed:
dvd_div_iff already exists

BoltonBailey committed:
Move lemmas to totient.lean

BoltonBailey committed:
fix line too long

BoltonBailey committed:
correct main file

BoltonBailey committed:
Add explanatory comments and golf

BoltonBailey committed:
space from code block

BoltonBailey committed:
rename variable

BoltonBailey committed:
more golfing

jcommelin commented on src/algebra/divisibility.lean:
```diff
@@ -102,6 +102,17 @@ alias dvd.intro_left ← dvd_of_mul_left_eq
+    { intro h,
+      cases h with c h',
+      exact dvd.intro_left c (eq.symm h'), },
+end
```
```suggestion
lemma dvd_iff_exists_eq_mul_left : a ∣ b ↔ ∃ c, b = c * a :=
⟨exists_eq_mul_left_of_dvd, by { rintro ⟨c, rfl⟩, exact ⟨c, mul_comm _ _⟩, }⟩
```

jcommelin commented on src/data/nat/totient.lean:
```diff
@@ -39,6 +39,25 @@ lemma totient_le (n : ℕ) : φ n ≤ n :=
+                  simp [pred_eq_sub_one n],
+                  simp [pos_of_gt hn],
+                end
+           ... < n : buffer.lt_aux_2 (pos_of_gt hn)
```
```suggestion
calc totient n ≤ ((range n).filter (≠ 0)).card :
  begin
    apply card_le_of_subset (monotone_filter_right _ _),
    intros n1 hn1 hn1',
    simpa only [hn1', coprime_zero_right, hn.ne'] using hn1,
  end
... = n - 1 : by simp only [filter_ne' (range n) 0, card_erase_of_mem, n.pred_eq_sub_one,
                card_range, pos_of_gt hn, mem_range]
... < n : buffer.lt_aux_2 (pos_of_gt hn)
```

jcommelin commented on src/data/nat/totient.lean:
```diff
@@ -194,6 +213,18 @@ begin
+begin
+  rw zmod.card_units_eq_totient p,
+  exact nat.le_pred_of_lt (nat.totient_lt p hp),
+end
```
```suggestion
lemma card_units_zmod_lt_sub_one {p : ℕ} (hp : 1 < p) [fact (0 < p)] :
  fintype.card (units (zmod p)) ≤ p - 1 :=
by { rw zmod.card_units_eq_totient p, exact nat.le_pred_of_lt (nat.totient_lt p hp), }
```

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,107 @@
+  (hr : 0 < r)
+  (ha : a^r = 1)
+  (hd : (∀ q : ℕ, (nat.prime q) -> (q ∣ r) -> a^(r/q) ≠ 1))
+  : order_of a = r :=
```
```suggestion
  (hr : 0 < r) (ha : a^r = 1) (hd : (∀ q : ℕ, (nat.prime q) -> (q ∣ r) -> a^(r/q) ≠ 1)) :
  order_of a = r :=
```

jcommelin reviewed (dismissed):
Please format according to the style guide: https://leanprover-community.github.io/contribute/style.html

BoltonBailey committed:
Update src/algebra/divisibility.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
Update src/data/nat/totient.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
Update src/data/nat/totient.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
style guide fix

jcommelin commented on src/data/nat/totient.lean:
```diff
@@ -194,6 +205,15 @@ begin
   rwa [succ_le_iff, pos_iff_ne_zero],
 end
 
+lemma card_units_zmod_lt_sub_one {p : ℕ} (hp : 1 < p) [fact (0 < p)] :
```
```suggestion
lemma card_units_zmod_lt_sub_one {p : ℕ} (hp : 1 < p)
  (hp' : fact (0 < p) := ⟨zero_lt_one.trans hp⟩) :
```
A little hack, to save TC some work.

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,103 @@
+This file implements the Lucas test for primes (not to be confused with the Lucas-Lehmer test for
+Mersenne primes). A number a witnesses that n is prime if a has order n-1 in the multiplicative
+group of integers mod n. This is checked by verifying that a^(n-1) = 1 (mod n) and a^d ≠ 1 (mod n)
+for any divisor d | n. This test is the basis of the Pratt primality certificate.
```
```suggestion
This file implements the Lucas test for primes (not to be confused with the Lucas-Lehmer test for
Mersenne primes). A number `a` witnesses that `n` is prime if `a` has order `n-1` in the multiplicative
group of integers mod `n`. This is checked by verifying that `a^(n-1) = 1 (mod n)` and `a^d ≠ 1 (mod n)`
for any divisor `d | n`. This test is the basis of the Pratt primality certificate.
```

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,103 @@
+variables (p : ℕ) [fact (0 < p)]
+
+/--
+If a^r = 1 mod p, but a^(r/q) ≠ 1 mod p for all prime factors q of r, then a has order r in the
```
Please put the maths between backticks `

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,103 @@
+end
+
+/--
+If a^(p-1) = 1 mod p, but a^((p-1)/q) ≠ 1 mod p for all prime factors q of p-1, then p is prime.
```
Idem.

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,103 @@
+-/
+theorem lucas_primality (a : zmod p)
+  (hp : 1 < p) (ha : a^(p-1) = 1)
+  (hd : (∀ q : ℕ, (nat.prime q) -> (q ∣ (p-1)) -> a^((p-1)/q) ≠ 1)) : p.prime :=
```
```suggestion
  (hd : ∀ q : ℕ, q.prime → q ∣ (p-1) → a^((p-1)/q) ≠ 1) : p.prime :=
```

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,103 @@
+have order p-1, which only happens when p is prime.
+-/
+theorem lucas_primality (a : zmod p)
+  (hp : 1 < p) (ha : a^(p-1) = 1)
```
This lemma is also true for `p = 1`, right? Because `hd` will be false (take `q = 2`).

BoltonBailey replied:
Well, that's one of these confusing things about zmod 1, it only works because 0 = 1. I did it this way because it seems like it would be confusing to people who come to the theorem later to reason about this.

BoltonBailey replied:
In my mind, it's more likely that someone would be trying to encounter problems thinking "Wait was it `0 < p` or `1 < p`, I guess its probably 1 < p because it doesn't make sense to compare something to 1 in zmod 1", than for someone to want to apply this theorem when p = 1, since the theorem is vacuous in that case anyway. 

If you feel strongly that the most general version of the theorem should be proved here then I can change it.

kbuzzard replied:
It might be because I'm an autistic pedant, but my feelings are the exact opposite to yours: I don't see any reason why it "doesn't make sense to compare something to 1 in zmod 1" -- are you using the "schoolchild" model of zmod n as the set {0,1,...,n-1} (so `n : zmod n` "has no meaning") as opposed to the quotient model (which Lean is using, so `n : zmod n` has a perfectly good meaning)? You should *always* state lemmas in maximal generality -- this is probably something which will become clear to you over time :-) (this was what happened to me, anyway)

kbuzzard replied:
Note however that the lemma isn't true for p=0 because a could be 1, unfortunately

Ruben-VandeVelde commented on src/data/nat/totient.lean:
```diff
@@ -39,6 +39,17 @@ lemma totient_le (n : ℕ) : φ n ≤ n :=
+  end
+... = n - 1 : by simp only [filter_ne' (range n) 0, card_erase_of_mem, n.pred_eq_sub_one,
+                card_range, pos_of_gt hn, mem_range]
+... < n : buffer.lt_aux_2 (pos_of_gt hn)
```
That doesn't sound like the lemma you're supposed to use. Maybe:
```suggestion
... < n : nat.pred_lt (pos_of_gt hn).ne'
```

BoltonBailey committed:
Update src/data/nat/totient.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
Backticks for math in comments

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
golf theorem statement

BoltonBailey committed:
Update src/data/nat/totient.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

BoltonBailey committed:
Add p = 1 case to theorem

BoltonBailey committed:
golf

BoltonBailey committed:
have from syntax

sgouezel commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,115 @@
+group must itself have order `p-1`, which only happens when `p` is prime.
+-/
+theorem lucas_primality (a : zmod p)
+  (hp : 0 < p) (ha : a^(p-1) = 1)
```
Can't you even completely remove the assumption `hp`? The fact that you have `a : zmod p` implies that `p` can not be zero.

sgouezel replied:
Oups, sorry, I thought that `zmod 0` was empty, and in fact it's Z. So forget this.

sgouezel replied:
In fact you can still exclude `p = 0` by using `hd`.

BoltonBailey replied:
Yes you can, I am doing this now

BoltonBailey replied:
Note, though, that `zmod p = \Z`, it's the hd condition that fails.

BoltonBailey committed:
deal with p = 0 case

BoltonBailey committed:
golfing

BoltonBailey committed:
Remove dependence on fact (0 < p)

BoltonBailey committed:
collapse arguments to line wrapping

BoltonBailey committed:
Update docstring

YaelDillies commented on src/data/nat/totient.lean:
```diff
@@ -39,6 +39,17 @@ lemma totient_le (n : ℕ) : φ n ≤ n :=
+  end
+... = n - 1 : by simp only [filter_ne' (range n) 0, card_erase_of_mem, n.pred_eq_sub_one,
+                card_range, pos_of_gt hn, mem_range]
+... < n : nat.pred_lt (pos_of_gt hn).ne'
```
Slightly API abusing golf.
```suggestion
... < n : nat.pred_lt hn.le
```

YaelDillies commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,102 @@
+This file implements the Lucas test for primes (not to be confused with the Lucas-Lehmer test for
+Mersenne primes). A number `a` witnesses that `n` is prime if `a` has order `n-1` in the
+multiplicative group of integers mod `n`. This is checked by verifying that `a^(n-1) = 1 (mod n)`
+and `a^d ≠ 1 (mod n)` for any divisor `d | n`. This test is the basis of the Pratt primality
```
```suggestion
and `a^d ≠ 1 (mod n)` for any divisor `d | n - 1`. This test is the basis of the Pratt primality
```

BoltonBailey committed:
Update src/data/nat/totient.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

BoltonBailey committed:
remove extra newlines

alexjbest commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,100 @@
+## TODO
+
+- Bonus: Show the reverse implication i.e. if a number is prime then it has a Lucas witness.
+  See [this stackexchange](https://math.stackexchange.com/a/59911/165144) answer for proof.
```
I'm confused what you intend for this TODO? We have the results https://leanprover-community.github.io/mathlib_docs/ring_theory/integral_domain.html#units.is_cyclic and even the same proof via https://leanprover-community.github.io/mathlib_docs/group_theory/specific_groups/cyclic.html#is_cyclic_of_card_pow_eq_one_le already.

BoltonBailey replied:
Cool, I wasn't aware the proof was in mathlib already. All I intend for the TODO then is to use this fact that the multiplicative group mod p is cyclic for prime p to show that there exists an `a` (namely a generator of that cyclic group), such that the test passes for `a` (in other words, that `∀ q : ℕ, q.prime → q ∣ (p-1) → a^((p-1)/q) ≠ 1`)

eric-wieser commented on src/data/nat/totient.lean:
```diff
@@ -194,6 +205,16 @@ begin
 
+lemma card_units_zmod_lt_sub_one {p : ℕ} (hp : 1 < p)
+  (hp' : fact (0 < p) := ⟨zero_lt_one.trans hp⟩) :
+  fintype.card (units (zmod p)) ≤ p - 1 :=
```
This assumption seems pretty weird to me; Perhaps it would be better to write
```suggestion
lemma card_units_zmod_lt_sub_one {p : ℕ} (hp : 1 < p) [fintype (units (zmod p))] :
  fintype.card (units (zmod p)) ≤ p - 1 :=
```

BoltonBailey replied:
@jcommelin was the one who wrote it this way, maybe he has an opinion. My understanding was that this way would be good, because when you're using it you don't have to write this typeclass argument when you use it.

jcommelin replied:
@BoltonBailey Probably the suggestion by Eric is better. If `fintype` didn't contain data, all of this would be unproblematic.

BoltonBailey replied:
I'm coming back to this now, and for whatever reason, it's not working this way, failing with "failed to synthesize typeclass instance". I have changed it back to the old way, not sure if some other change in mathlib caused it to stop working - did it work this way a month ago?

eric-wieser commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,100 @@
+      exact hd 2 nat.prime_two (by simp only [h0, zero_sub', dvd_zero]) (by simp [h0]), },
+  by_cases h1 : p = 1,
+    { exfalso,
+      exact hd 2 nat.prime_two (by simp only [h1, nat.sub_self, dvd_zero]) (by simp [h1]), },
```
I think this would be better as:
```suggestion
  have h0 : p ≠ 0,
  { intro h0,
    exact hd 2 nat.prime_two (by simp only [h0, zero_sub', dvd_zero]) (by simp [h0]), },
  have h1 : p ≠ 1,
  { intro h1,
    exact hd 2 nat.prime_two (by simp only [h1, nat.sub_self, dvd_zero]) (by simp [h1]), },
```

YaelDillies replied:
Or even
```suggestion
  have h0 : p ≠ 0,
  { exact λ h0, hd 2 nat.prime_two (by simp only [h0, zero_sub', dvd_zero]) (by simp [h0]) },
  have h1 : p ≠ 1,
  { exact λ h1, hd 2 nat.prime_two (by simp only [h1, nat.sub_self, dvd_zero]) (by simp [h1]) },
```

BoltonBailey committed:
Note mathlib theorem in TODO

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

BoltonBailey committed:
Update src/data/nat/totient.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

BoltonBailey committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into BoltonBailey/lucas-primality

BoltonBailey committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into BoltonBailey/lucas-primality

BoltonBailey committed:
bugfix

BoltonBailey committed:
move theorem to monoid order file

BoltonBailey committed:
fix main theorem

BoltonBailey committed:
rename variables

BoltonBailey committed:
more renaming variables

BoltonBailey committed:
rename lemma

BoltonBailey committed:
fix bug

BoltonBailey committed:
fix line

eric-wieser committed:
remove [fact (0 < n)] where it is not needed

feel free to revert, github would not allow me to make suggestions on the relevant lines

BoltonBailey committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into BoltonBailey/lucas-primality

BoltonBailey committed:
attempt to add to_additive

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,70 @@
+  have hp1 : 1 < p := lt_of_le_of_ne (nat.succ_le_iff.mpr (zero_lt_iff.mpr h0)) (ne.symm h1),
+  have order_of_a : order_of a = p-1,
+    { apply order_of_eq_of_pow_and_pow_div_prime _ ha hd,
+      exact tsub_pos_of_lt hp1, },
```
```suggestion
  { apply order_of_eq_of_pow_and_pow_div_prime _ ha hd,
    exact tsub_pos_of_lt hp1, },
```

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,70 @@
+  have order_of_a : order_of a = p-1,
+    { apply order_of_eq_of_pow_and_pow_div_prime _ ha hd,
+      exact tsub_pos_of_lt hp1, },
+  haveI fhp0 : fact (0 < p) := {out := zero_lt_iff.mpr h0},
```
```suggestion
  haveI fhp0 : fact (0 < p) := ⟨zero_lt_iff.mpr h0⟩,
```

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,70 @@
+  -- Prove cardinality of `units` of `zmod p` is both `≤ p-1` and `≥ p-1`
+  rw le_antisymm_iff,
+  split,
+    { exact nat.card_units_zmod_lt_sub_one hp1, },
```
Please unindent the rest of the proof

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,70 @@
+  split,
+    { exact nat.card_units_zmod_lt_sub_one hp1, },
+    { have hp' : p - 2 + 1 = p - 1,
+        { apply eq.symm,
```
This subproof is also indented too much

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,70 @@
+          linarith, },
+      let a' : units (zmod p) := units.mk_of_mul_eq_one a (a ^ (p-2)) (by rw [←pow_succ, hp', ha]),
+      have a_coe : a = units.coe_hom (zmod p) a',
+        { unfold_coes, simp, },
```
Idem

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,70 @@
+      have a_coe : a = units.coe_hom (zmod p) a',
+        { unfold_coes, simp, },
+      have order_of_a' : order_of a' = p-1,
+        { rw [←order_of_a, a_coe, order_of_injective (@units.coe_hom (zmod p) _)],
```
Idem

jcommelin commented on src/group_theory/order_of_element.lean:
```diff
@@ -161,6 +161,35 @@ begin
+  -- Assume `a` is not one...
+  by_contra,
+  have a_min_fac_dvd_p_sub_one : a.min_fac ∣ n,
+    { have b_dvd_factor : ∃ (b : ℕ), a = b * a.min_fac,
```
Please unindent this block

jcommelin commented on src/data/nat/totient.lean:
```diff
@@ -39,14 +39,31 @@ lemma totient_le (n : ℕ) : φ n ≤ n :=
+  apply nat.pred_lt,
+  rw nat.sub,
+  linarith,
+end
```
```suggestion
... < n : nat.sub_lt (pos_of_gt hn) zero_lt_one
```

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
unindent blocks too indented

BoltonBailey committed:
unindent block

BoltonBailey committed:
fix have block

BoltonBailey committed:
Update src/data/nat/totient.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
typo in header

BoltonBailey committed:
modify docstring

BoltonBailey committed:
reline args for clarity

BoltonBailey committed:
renamed to_additive lemma

BoltonBailey committed:
repluralize authors

jcommelin commented on src/group_theory/order_of_element.lean:
```diff
@@ -161,6 +161,34 @@ begin
+  by_contra,
+  have a_min_fac_dvd_p_sub_one : a.min_fac ∣ n,
+  { have b_dvd_factor : ∃ (b : ℕ), a = b * a.min_fac := exists_eq_mul_left_of_dvd a.min_fac_dvd,
+    cases b_dvd_factor with b hb,
```
```suggestion
  { obtain ⟨b, hb⟩ : ∃ (b : ℕ), a = b * a.min_fac := exists_eq_mul_left_of_dvd a.min_fac_dvd,
```

jcommelin commented on src/group_theory/order_of_element.lean:
```diff
@@ -161,6 +161,34 @@ begin
+  { have b_dvd_factor : ∃ (b : ℕ), a = b * a.min_fac := exists_eq_mul_left_of_dvd a.min_fac_dvd,
+    cases b_dvd_factor with b hb,
+    rw [hb, ←mul_assoc] at ha,
+    exact dvd.intro_left (order_of x * b) (eq.symm ha), },
```
```suggestion
    exact dvd.intro_left (order_of x * b) ha.symm, },
```

jcommelin commented on src/group_theory/order_of_element.lean:
```diff
@@ -161,6 +161,34 @@ begin
+      ha, mul_comm, nat.mul_dvd_mul_iff_left (order_of_pos' _)],
+  { exact nat.min_fac_dvd a, },
+  { rw is_of_fin_order_iff_pow_eq_one,
+    exact Exists.intro n (id ⟨hn, hx⟩)},
```
```suggestion
    exact Exists.intro n (id ⟨hn, hx⟩) },
```

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,70 @@
+    { apply eq.symm,
+      rw [nat.sub_eq_iff_eq_add, add_assoc, one_add_one_eq_two, ←nat.sub_eq_iff_eq_add],
+      linarith,
+      linarith, },
```
This suggests to me that there are multiple goals. Either use `{..}` to structure the subproofs, or use `; linarith` to execute the tactic on both goals.

BoltonBailey replied:
`zify` actually seems like a good way to resolve this, so I went that route.

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,70 @@
+  split,
+  { exact nat.card_units_zmod_lt_sub_one hp1, },
+  { have hp' : p - 2 + 1 = p - 1,
+    { apply eq.symm,
```
This can be done using the `symmetry` tactic, which might be slightly more idiomatic.

BoltonBailey replied:
I golfed this using zify, hope that's ok too.

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,70 @@
+  -- Prove cardinality of `units` of `zmod p` is both `≤ p-1` and `≥ p-1`
+  rw le_antisymm_iff,
+  split,
+  { exact nat.card_units_zmod_lt_sub_one hp1, },
```
If you want, you can golf this using
```suggestion
  refine ⟨nat.card_units_zmod_lt_sub_one hp1, _⟩,
```
and then you can unindent the second block, and remove the `{..}`.

BoltonBailey committed:
Update src/group_theory/order_of_element.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
Update src/group_theory/order_of_element.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
Update src/group_theory/order_of_element.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

BoltonBailey committed:
unwrap and unindent

BoltonBailey committed:
golf using zify

jcommelin commented on src/number_theory/lucas_primality.lean:
```diff
@@ -0,0 +1,68 @@
+  { rw [←order_of_a, a_coe, order_of_injective (@units.coe_hom (zmod p) _)],
+    exact units.ext, },
+  rw ←order_of_a',
+  apply order_of_le_card_univ,
```
```suggestion
  have h0 : p ≠ 0, { rintro ⟨⟩, exact hd 2 nat.prime_two (dvd_zero _) (pow_zero _) },
  have h1 : p ≠ 1, { rintro ⟨⟩, exact hd 2 nat.prime_two (dvd_zero _) (pow_zero _) },
  have hp1 : 1 < p := lt_of_le_of_ne h0.bot_lt h1.symm,
  have order_of_a : order_of a = p-1,
  { apply order_of_eq_of_pow_and_pow_div_prime _ ha hd,
    exact tsub_pos_of_lt hp1, },
  haveI fhp0 : fact (0 < p) := ⟨h0.bot_lt⟩,
  rw nat.prime_iff_card_units,
  -- Prove cardinality of `units` of `zmod p` is both `≤ p-1` and `≥ p-1`
  refine le_antisymm (nat.card_units_zmod_lt_sub_one hp1) _,
  have hp' : p - 2 + 1 = p - 1 := tsub_add_eq_add_tsub hp1,
  let a' : units (zmod p) := units.mk_of_mul_eq_one a (a ^ (p-2)) (by rw [←pow_succ, hp', ha]),
  calc p - 1 = order_of a : order_of_a.symm
  ... = order_of a' : order_of_injective (units.coe_hom (zmod p)) units.ext a'
  ... ≤ fintype.card (units (zmod p)) : order_of_le_card_univ,
```

BoltonBailey committed:
Update src/number_theory/lucas_primality.lean

Co-authored-by: Johan Commelin <johan@commelin.net>
